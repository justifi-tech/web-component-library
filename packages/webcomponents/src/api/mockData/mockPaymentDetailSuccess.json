{
  "id": "py_1NNeEnf4FbelxDCQN2RHcE",
  "type": "payment",
  "page_info": null,
  "data": {
    "id": "py_1NNeEnf4FbelxDCQN2RHcE",
    "account_id": "acc_5Et9iXrSSAZR2KSouQGAWi",
    "amount_disputed": 0,
    "amount_refunded": 0,
    "amount_returned": 0,
    "amount": 23456,
    "amount_refundable": 23456,
    "application_fee_rate_id": null,
    "balance": 23256,
    "capture_strategy": "automatic",
    "captured": true,
    "created_at": "2023-12-27T21:52:53.286Z",
    "currency": "usd",
    "description": "payment",
    "statement_descriptor": "WC Account",
    "disputed": false,
    "error_code": null,
    "error_description": null,
    "fee_amount": 200,
    "financial_transaction_id": "ft_1czgEJzmHBw7AHMTAeJtEk",
    "is_test": true,
    "metadata": null,
    "payment_intent_id": null,
    "refunded": false,
    "returned": false,
    "status": "succeeded",
    "updated_at": "2023-12-27T21:53:18.276Z",
    "payment_method": {
      "bank_account": {
        "id": "pm_uICmv8696AatL4gRwa0Wq",
        "acct_last_four": "6789",
        "name": "Julia Garnett",
        "brand": "Altra",
        "token": "pm_uICmv8696AatL4gRwa0Wq",
        "created_at": "2023-12-27T21:52:53.277Z",
        "updated_at": "2023-12-27T21:52:53.277Z"
      },
      "customer_id": "cust_4g8yPTP4hjeid0tPuqpk65",
      "signature": "hOl9oDjAd7bP9qCESVSsV"
    },
    "application_fee": {
      "id": "fee_2BBieeBBZRkEPOOMKJAAAI",
      "amount": 200,
      "currency": "usd",
      "created_at": "2023-12-27T21:52:53.444Z",
      "updated_at": "2023-12-27T21:53:18.270Z"
    },
    "refunds": [],
    "disputes": [],
    "transaction_hold": null
  }
}
