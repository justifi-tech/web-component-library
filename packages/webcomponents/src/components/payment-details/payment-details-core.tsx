import { Component, Host, h, Prop, State, Watch } from '@stencil/core';
import { Payment } from '../../api';
import { MapPaymentStatusToBadge, formatCurrency, formatDate, formatTime, snakeCaseToHumanReadable } from '../../utils/utils';
import { CodeBlock, DetailItem, DetailSection, EntityHeadInfo, EntityHeadInfoItem, ErrorState, LoadingState } from '../details/utils';
import { PaymentService } from '../payments-list/payment.service';

@Component({
  tag: 'payment-details-core',
  styleUrl: 'payment-details.scss',
})

export class PaymentDetailsCore {
  @Prop() paymentId: string;
  @Prop() authToken: string;
  @Prop() paymentService: PaymentService;
  @State() payment: Payment;
  @State() loading: boolean = true;
  @State() errorMessage: string;

  @Watch('paymentId')
  @Watch('authToken')
  @Watch('paymentService')
  updateOnPropChange() {
    this.fetchData();
  }

  connectedCallback() {
    this.fetchData();
  }

  async fetchData(): Promise<void> {
    if (!this.paymentId || !this.authToken) {
      this.errorMessage = "Can not fetch any data without a PaymentID and an AuthToken";
      this.loading = false;
      return;
    }

    if (!this.paymentService) {
      return;
    }

    this.loading = true;

    try {
      const response = await this.paymentService.fetchPayment(this.paymentId, this.authToken);

      if (!response.error) {
        this.payment = new Payment(response.data)
        this.errorMessage = null;
      } else {
        const responseError = typeof response.error === 'string' ? response.error : response.error.message;
        this.errorMessage = `Error trying to fetch data : ${responseError}`;
        console.error(this.errorMessage);
      }
    } catch (error) {
      this.errorMessage = `Error trying to fetch data : ${error}`;
      console.error(this.errorMessage);
    } finally {
      this.loading = false;
    }
  }

  render() {
    return (
      <Host>
        {this.loading && LoadingState}
        {!this.loading && this.errorMessage && ErrorState(this.errorMessage)}
        {!this.loading && !this.errorMessage &&
          this.payment && (
            <justifi-details error-message={this.errorMessage}>
              <EntityHeadInfo
                slot="head-info"
                badge={<span slot='badge' innerHTML={MapPaymentStatusToBadge(this.payment.status)} />}
                title={`${formatCurrency(this.payment.amount)} ${this.payment?.currency.toUpperCase()}`}
              >
                <EntityHeadInfoItem
                  classes="border-1 border-end"
                  title="Updated At"
                  value={`${formatDate(this.payment.updated_at)} ${formatTime(this.payment.updated_at)}`}
                />
                <EntityHeadInfoItem
                  classes="border-1 border-end"
                  title="Created At"
                  value={`${formatDate(this.payment.created_at)} ${formatTime(this.payment.created_at)}`}
                />
                <EntityHeadInfoItem title="ID" value={this.payment.id} />
              </EntityHeadInfo>
              <div slot='detail-sections'>
                <DetailSection sectionTitle="Details">
                  <DetailItem title="Amount" value={formatCurrency(this.payment.amount)} />
                  <DetailItem title="Fees" value={formatCurrency(this.payment.fee_amount)} />
                  <DetailItem title="Refunded" value={formatCurrency(this.payment.amount_refunded)} />
                  <DetailItem title="Net" value={formatCurrency(this.payment.balance)} />
                  <DetailItem title="Status" value={MapPaymentStatusToBadge(this.payment.status)} />
                  <DetailItem title="Payment ID" value={this.payment.id} />
                  <DetailItem title="Processing Fees" value={formatCurrency(this.payment.fee_amount)} />
                  <DetailItem title="Statement Descriptor" value={this.payment.statement_descriptor} />
                  <DetailItem title="Description" value={this.payment.description} />
                </DetailSection>
                {this.payment.payment_method.card && (
                  <DetailSection sectionTitle="Payment Method">
                    <DetailItem title="ID" value={this.payment.payment_method.card.id} />
                    <DetailItem title="Payment Type" value="Card" />
                    <DetailItem title="Last 4 Numbers" value={this.payment.payment_method.lastFourDigits} />
                    <DetailItem title="Brand" value={snakeCaseToHumanReadable(this.payment.payment_method.card.brand)} />
                    <DetailItem title="Cardholder" value={this.payment.payment_method.payersName} />
                  </DetailSection>
                )}
                {this.payment.payment_method.bank_account && (
                  <DetailSection sectionTitle="Payment Method">
                    <DetailItem title="ID" value={this.payment.payment_method.bank_account.id} />
                    <DetailItem title="Last 4 Numbers" value={this.payment.payment_method.lastFourDigits} />
                    <DetailItem title="Bank Name" value={this.payment.payment_method.bank_account.brand} />
                    <DetailItem title="Account Owner" value={this.payment.payment_method.payersName} />
                  </DetailSection>
                )}
                <DetailSection sectionTitle='Metadata'>
                  <CodeBlock metadata={this.payment.metadata} />
                </DetailSection>
              </div>
            </justifi-details>
          )}
      </Host>
    );
  }
}
