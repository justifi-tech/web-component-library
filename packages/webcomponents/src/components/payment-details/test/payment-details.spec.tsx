import { newSpecPage } from '@stencil/core/testing';
import { PaymentDetailsCore } from '../payment-details-core';
import mockPaymentDetailSuccess from '../../../api/mockData/mockPaymentDetailSuccess.json';

describe('payment-details-core', () => {
  let paymentServiceMock;

  beforeEach(() => {
    paymentServiceMock = {
      fetchPayment: jest.fn(),
    };
  });

  it('renders loading state initially', async () => {
    const page = await newSpecPage({
      components: [PaymentDetailsCore],
      html: '<payment-details-core></payment-details-core>',
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders error state when no paymentId or authToken', async () => {
    const page = await newSpecPage({
      components: [PaymentDetailsCore],
      html: '<payment-details-core></payment-details-core>',
    });
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('fetches payment details on connected callback', async () => {
    paymentServiceMock.fetchPayment.mockResolvedValueOnce(mockPaymentDetailSuccess);
    const page = await newSpecPage({
      components: [PaymentDetailsCore],
      html: '<payment-details-core></payment-details-core>',
      supportsShadowDom: true,
    });

    page.rootInstance.paymentId = '123';
    page.rootInstance.authToken = 'token';
    page.rootInstance.paymentService = paymentServiceMock;
    await page.waitForChanges();

    expect(paymentServiceMock.fetchPayment).toHaveBeenCalledWith('123', 'token');
    expect(page.rootInstance.payment).toEqual(expect.objectContaining({
      "id": "py_1NNeEnf4FbelxDCQN2RHcE"
    }));
    expect(page.root).toMatchSnapshot();
  });

  it('handles fetch payment error correctly', async () => {
    paymentServiceMock.fetchPayment.mockRejectedValueOnce('Fetch error');
    const page = await newSpecPage({
      components: [PaymentDetailsCore],
      html: '<payment-details-core></payment-details-core>',
    });

    page.root.paymentId = '123';
    page.root.authToken = 'token';
    page.root.paymentService = paymentServiceMock;
    await page.waitForChanges();

    expect(page.rootInstance.errorMessage).toBe('Error trying to fetch data : Fetch error');
    expect(page.root).toMatchSnapshot();
  });
});

