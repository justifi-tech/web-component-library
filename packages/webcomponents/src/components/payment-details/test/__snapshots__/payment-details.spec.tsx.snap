// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`payment-details-core fetches payment details on connected callback 1`] = `
<payment-details-core>
  <justifi-details>
    <div class="p-2" part="detail-head" slot="head-info">
      <div class="align-items-center d-flex flex-row gap-2 mb-2">
        <h1 class="m-0" part="detail-title">
          $234.56 USD
        </h1>
        <span slot="badge">
          <span class="badge bg-success" title="This payment was successfully captured">
            Successful
          </span>
        </span>
      </div>
      <div class="align-items-top d-flex flex-row" part="detail-head-info">
        <div class="border-1 border-end d-flex flex-column" part="detail-head-info-item">
          <span class="border-1 border-bottom fw-bold pe-2 ps-2" part="detail-head-info-item-title">
            Updated At
          </span>
          <span class="pe-2 ps-2 pt-2" part="detail-head-info-item-data">
            Dec 27, 2023 6:53pm
          </span>
        </div>
        <div class="border-1 border-end d-flex flex-column" part="detail-head-info-item">
          <span class="border-1 border-bottom fw-bold pe-2 ps-2" part="detail-head-info-item-title">
            Created At
          </span>
          <span class="pe-2 ps-2 pt-2" part="detail-head-info-item-data">
            Dec 27, 2023 6:52pm
          </span>
        </div>
        <div class="d-flex flex-column undefined" part="detail-head-info-item">
          <span class="border-1 border-bottom fw-bold pe-2 ps-2" part="detail-head-info-item-title">
            ID
          </span>
          <span class="pe-2 ps-2 pt-2" part="detail-head-info-item-data">
            py_1NNeEnf4FbelxDCQN2RHcE
          </span>
        </div>
      </div>
    </div>
    <div slot="detail-sections">
      <div class="mt-4" part="detail-section">
        <h5 part="detail-section-title">
          Details
        </h5>
        <hr>
        <div class="d-table gap-2 w-100">
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
              Amount
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
              $234.56
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
              Fees
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
              $2.00
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
              Refunded
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
              $0.00
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
              Net
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
              $232.56
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
              Status
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
              <span class="badge bg-success" title="This payment was successfully captured">
                Successful
              </span>
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
              Payment ID
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
              py_1NNeEnf4FbelxDCQN2RHcE
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
              Processing Fees
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
              $2.00
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
              Description
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
              payment
            </span>
          </div>
        </div>
      </div>
      <div class="mt-4" part="detail-section">
        <h5 part="detail-section-title">
          Payment Method
        </h5>
        <hr>
        <div class="d-table gap-2 w-100">
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
              ID
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
              pm_uICmv8696AatL4gRwa0Wq
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
              Last 4 Numbers
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
              **** 6789
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
              Bank Name
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
              Altra
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="detail-section-item-title">
              Account Owner
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="detail-section-item-data">
              Julia Garnett
            </span>
          </div>
        </div>
      </div>
      <div class="mt-4" part="detail-section">
        <h5 part="detail-section-title">
          Metadata
        </h5>
        <hr>
        <div class="d-table gap-2 w-100"></div>
      </div>
    </div>
  </justifi-details>
</payment-details-core>
`;

exports[`payment-details-core handles fetch payment error correctly 1`] = `
<payment-details-core>
  <main class="d-flex justify-content-center p-4 text-center" part="detail-empty-state" style="font-size: 1.2rem;">
    <div>
      Error trying to fetch data : Fetch error
    </div>
  </main>
</payment-details-core>
`;

exports[`payment-details-core renders error state when no paymentId or authToken 1`] = `
<payment-details-core>
  <main class="d-flex justify-content-center p-4 text-center" part="detail-empty-state" style="font-size: 1.2rem;">
    <div>
      Can not fetch any data without a PaymentID and an AuthToken
    </div>
  </main>
</payment-details-core>
`;

exports[`payment-details-core renders loading state initially 1`] = `
<payment-details-core>
  <main class="d-flex justify-content-center p-4 text-center" part="detail-empty-state" style="font-size: 1.2rem;">
    <div>
      Can not fetch any data without a PaymentID and an AuthToken
    </div>
  </main>
</payment-details-core>
`;
